Issues/4093 swap lists

aerych opened the pull request:
This PR adds the final touches on the refreshed reader post list and cards. 
Closes #4093, #4032
For comparison the prototypes are here: https://marvelapp.com/66gi8f#7304611 

Changes include:
- The old list is deleted.
- The new post list is used everywhere the old one was previously used. 
- Headers for tags, lists and sites are no longer placeholders
- Following/Unfollowing from tag and site headers is enabled
- Tapping a tag will preview that tags posts in a new list.
- The logged out reader does not show/enable features that require a login, e.g. likes, blocking, following.

To test:
Menu:
- Test the menu when logged out and when logged in. 
- The first section is default topics and has not title. There should be a section for tags, and may or may not be a section for lists depending on if you're following any. 
- Confirm following a topic still works via the search bar.
- Confirm following a site still works via the search bar. 
- Confirm tags can still be unfollowed via the menu
- Confirm sites can still be unfollowed via the menu

Default topics (freshly pressed, blogs I follow)
- Ensure these topics do not show a header.

Tags
- Ensure tags show a tag header with the tag name and follow button. 
- Test following / unfollowing via the header. 

Lists
- Ensure lists show a list header with the name of the list. 

Sites
- Ensure sites show a header that includes the blavatar, site name, url, subscriber count and a description (if there is one). 
- Test following / unfollowing via the header.

General:
- Test pull to refresh & infinite scroll
- Tap the name of a site in a post card to preview its site. Confirm that the site name is disabled in the cards for the preview
- Tap to preview a tag.
- Confirm you can like and unlike a post
- Confirm you can view comments for a post
- Confirm you can block a site via the ellipsis menu
- Confirm you can share a post via the ellipsis menu


@koke would you be up for a review of this one? 

Needs Review: @koke

koke commented:
Testing after upgrading from `develop`. When I tap the topic selector, I get a crash on `ReaderStreamViewController.predicateForFetchRequest()`:

> fatal error: unexpectedly found nil while unwrapping an Optional value

It seems it fails to find the default topic:

```managedObjectContext().existingObjectWithID(readerTopic!.objectID)``` throws an error, `topic` stays uninitialized, and `NSPredicate(format: "topic = %@ AND isSiteBlocked = NO", topic)` crashes

This might be helpful :wink: 

```
(lldb) po readerTopic!.managedObjectContext!
<NSManagedObjectContext: 0x12e5980f0>

(lldb) po managedObjectContext()
<NSManagedObjectContext: 0x13000f1f0>

(lldb) po readerTopic!.managedObjectContext!
<NSManagedObjectContext: 0x12e5980f0>
```

koke commented:
More info:

```
(lldb) po managedObjectContext().existingObjectWithID(readerTopic!.objectID)
Error Domain=NSCocoaErrorDomain Code=133000 "Se ha intentado acceder a un objeto no encontrado en el almacenamiento."

(lldb) po readerTopic
▿ Optional(<WordPress.ReaderDefaultTopic: 0x12929b390> (entity: ReaderDefaultTopic; id: 0x12929b850 <x-coredata:///ReaderDefaultTopic/t74554040-CD90-4FE6-B8A6-EEF224E2D8CF8> ; data: {
    following = 1;
    lastSynced = nil;
    path = "https://public-api.wordpress.com/rest/v1.2/freshly-pressed";
    posts =     (
    );
    showInMenu = 1;
    title = "Freshly Pressed";
    type = default;
}))

(lldb) po (managedObjectContext().executeFetchRequest(NSFetchRequest(entityName: "ReaderAbstractTopic")) as! [ReaderAbstractTopic]).filter { return $0.path == readerTopic!.path }
▿ 1 elements
  - [0] : <WordPress.ReaderDefaultTopic: 0x129728fb0> (entity: ReaderDefaultTopic; id: 0x12929b850 <x-coredata:///ReaderDefaultTopic/t74554040-CD90-4FE6-B8A6-EEF224E2D8CF8> ; data: {
    following = 1;
    lastSynced = nil;
    path = "https://public-api.wordpress.com/rest/v1.2/freshly-pressed";
    posts = "<relationship fault: 0x1296b2fb0 'posts'>";
    showInMenu = 1;
    title = "Freshly Pressed";
    type = default;
})

(lldb) po managedObjectContext().existingObjectWithID(readerTopic!.objectID)
<WordPress.ReaderDefaultTopic: 0x129728fb0> (entity: ReaderDefaultTopic; id: 0x12929b850 <x-coredata:///ReaderDefaultTopic/t74554040-CD90-4FE6-B8A6-EEF224E2D8CF8> ; data: {
    following = 1;
    lastSynced = nil;
    path = "https://public-api.wordpress.com/rest/v1.2/freshly-pressed";
    posts = "<relationship fault: 0x1296b2fb0 'posts'>";
    showInMenu = 1;
    title = "Freshly Pressed";
    type = default;
})
```

That's really odd. `existingObjectWithID` fails to fetch the objectID, but it works after requesting all topics. I'm wondering if something changed for core data on iOS9. @astralbodies any ideas?

koke commented:
Not really new code, but I found this while debugging the crash. In `-[ReaderTopicService mergeMenuTopics:withSuccess:]`:

```objc
            for (ReaderAbstractTopic *topic in currentTopics) {
                if (![topic isKindOfClass:[ReaderSiteTopic class]] && ![topicsToKeep containsObject:topic]) {
                    DDLogInfo(@"Deleting Reader Topic: %@", topic);
                    if ([topic isEqual:self.currentTopic]) {
                        self.currentTopic = nil;
                    }
```

It took me a while to figure out why it was deleting the existing default topic every time, until I noticed the version change in `path`:

```
(lldb) po topic
<WordPress.ReaderDefaultTopic: 0x1457c2760> (entity: ReaderDefaultTopic; id: 0xd00000000008000c <x-coredata://71885D61-3E47-48E2-BD28-C7DC5FA8AAA9/ReaderDefaultTopic/p2> ; data: {
    following = 0;
    lastSynced = "2015-09-17 09:08:46 +0000";
    path = "https://public-api.wordpress.com/rest/v1.1/freshly-pressed";
    posts = "<relationship fault: 0x146b7fa50 'posts'>";
    showInMenu = 1;
    title = "Freshly Pressed";
    type = default;
})

(lldb) po [[topicsToKeep filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"title = 'Freshly Pressed'"]] firstObject]
<WordPress.ReaderDefaultTopic: 0x146b82570> (entity: ReaderDefaultTopic; id: 0x146bf0810 <x-coredata:///ReaderDefaultTopic/t55B411B5-E07C-4F04-889B-A6FA360B08858> ; data: {
    following = 1;
    lastSynced = nil;
    path = "https://public-api.wordpress.com/rest/v1.2/freshly-pressed";
    posts =     (
    );
    showInMenu = 1;
    title = "Freshly Pressed";
    type = default;
})
```

koke commented:
Another unexpected, non-new-code thing, and why the managed context don't match. On `ReaderStreamViewController`:

```swift
public func managedObjectContext() -> NSManagedObjectContext {
    if let context = displayContext {
        return context
    }
    displayContext = NSManagedObjectContext(concurrencyType: .MainQueueConcurrencyType)
    displayContext!.parentContext = ContextManager.sharedInstance().mainContext
    return displayContext!
}
```

First, I'm not sure if this has to be this way because it's `@objc`, but the _swifty_ way of doing this would be dropping the `displayContext` var, and turn `managedObjectContext` into a lazy property: 

```swift
lazy var managedObjectContext: NSManagedObjectContext = {
    let displayContext = NSManagedObjectContext(concurrencyType: .MainQueueConcurrencyType)
    displayContext!.parentContext = ContextManager.sharedInstance().mainContext
    return displayContext!
}()
```

But the relevant issue here is:
1. Not using `ContextManager` to acquire a context
2. Using a main queue context other than `mainContext`, and make it a child of `mainContext`

While 2 might be a valid pattern (I see from #4225 that it was intentional), what's happening here is:

1. `ReaderSubscriptionViewController` calls `syncTopics` on `mainContext`
2. New topics arrive and are merged, the merge method notices our current topic is no longer valid (because of the version change), and sets `currentTopic = nil`
3. This triggers a `ReaderTopicDidChangeNotification`
4. `ReaderViewController` gets the notification and sets `postsViewController.readerTopic = topic`
5. `ReaderStreamViewController` tries to find a topic that matches readerTopic on its child context and fails

Since the stream view controller uses a child context, I don't think it has the new topics yet. I guess somehow doing the fetch request or using `objectWithID` makes the child context fetch from the parent.

```
(lldb) po managedObjectContext().existingObjectWithID(readerTopic!.objectID)
Error Domain=NSCocoaErrorDomain Code=133000 "Se ha intentado acceder a un objeto no encontrado en el almacenamiento."

(lldb) po displayContext!.registeredObjects
0 elements

(lldb) po managedObjectContext().objectWithID(readerTopic!.objectID)
<WordPress.ReaderDefaultTopic: 0x146be18c0> (entity: ReaderDefaultTopic; id: 0x146b36010 <x-coredata:///ReaderDefaultTopic/t4617701E-3D0E-40CD-9EC3-5A600C2ACF608> ; data: <fault>)

(lldb) po displayContext!.registeredObjects
▿ 1 elements
  - [0] : <WordPress.ReaderDefaultTopic: 0x146be18c0> (entity: ReaderDefaultTopic; id: 0x146b36010 <x-coredata:///ReaderDefaultTopic/t4617701E-3D0E-40CD-9EC3-5A600C2ACF608> ; data: <fault>)
```

koke commented:
Site streams
-----------------

When you tap on a site, the site icon is missing on the header view:

![site-icon-missing](https://cloud.githubusercontent.com/assets/8739/9930847/07863f76-5d37-11e5-945d-93ea4d2c96bb.png)

Also, it would be nice to show the header while it's loading. The transition feels a bit overwhelming when you add up the push animation, 'Loading stream', 'Fetching Posts' with it's own animation, and finally showing posts.

![site-loading-progression](https://cloud.githubusercontent.com/assets/8739/9931014/1e2b8762-5d38-11e5-9492-d587ddcb927d.png)

https://cloudup.com/cvp20Nv9AyK

It also seems like no data is cached when you visit a site a second time?

Stream layout
-------------------

When you compare the PR with the prototype you linked, there are a few discrepancies:

![marvel-vs-pr](https://cloud.githubusercontent.com/assets/8739/9931240/427f453e-5d3a-11e5-99b4-2f21ef9157bf.png)

Overall, the text on the PR feels more *packed* and needs some breathing space. In detail:
- Title font looks slightly smaller and line height doesn't match
- Post content is too close to the footer. The prototype has the *X words* line in between, so I can't compare directly
- Site title color doesn't match
- Author/timestamp seem to use a lighter font and/or a lighter gray. 
- Bottom margin of the gravatar is slightly larger than top margin. I believe it's the image which is smaller than it should be.

I'll summon @shaunandrews for a design review :)

shaunandrews commented:
Nice eye, @koke. You've hit pretty much everything. @aerych, here's a quick blueprint I threw together that outlines the spacing in the prototype:

![ios reader card blueprint](https://cloud.githubusercontent.com/assets/191598/9935449/ae4b1024-5d24-11e5-87f7-cbbb2caa9512.png)

I'm not super familiar with any grid spec's we may have for iOS. I dug around a bit but couldn't find anything concrete.

aerych commented:
Quick summary of chatting with @shaunandrews.   
For layout consistency within the app, we'll let much of the horizontal spacing and vertical spacing in the header in the reader post cards match what we have for the cards in the my sites post list. See #4285
We'll also keep the blue for the site name when its tappable.

aerych commented:
>  it would be nice to show the header while it's loading

I'd really love to do this, and had originally written the feature that way.  The gotcha was site headers with very long descriptions that overlap the loading animation.  Its especially bad in landscape orientation.  For now, no headers while loading, but I'd happy revisit this when we can iterate on the loading screens in the app.

aerych commented:
> I'm not sure if this has to be this way because it's @objc, but the swifty way of doing this would be dropping the displayContext var, and turn managedObjectContext into a lazy property

The method is defined as part of the `WPTableViewHandlerDelegate` protocol otherwise the lazy var would be an option.

> Not using ContextManager to acquire a context

#4263 is a reminder to add a convenience method to `ContextManager` to do this sort of thing.

aerych commented:
@koke :  Thanks for the excellent debugging on the topic version. :)  I've added a patch to ensure the context is always current when configuring a new topic, and an early return in case of an error as a safety net.  I'll be glad when there are no longer any logic paths that end up recycling a post list, but that must wait til the menu refactor.  Also the blavatar should be loading correctly now on the site detail.  

@shaunandrews :  I've updated layout to what I *think* matches the grid. (I went ahead and updated to the 16px grid after all.)  There is a bit of math happening to account for the differences between line height, and text baselines and ascenders. Please double check to make sure I didn't miss any.  To make it easier I've created gallery of screen shots here: https://cloudup.com/cXo7mKiqS0Z
One thing of note... the iPad uses larger fonts and line heights for the title and summary.  This is consistent with what we're doing in the My Sites post list.  The sizes are:
- title: 24 font, 32 line
- content: 16 font: 26 line

Let me know if we need to make a change.

shaunandrews commented:
Could we drop color down a shade to #C8D7E1 for these lines around the follower count on both iPhone and iPad:

![screen shot 2015-09-17 at 11 37 08 pm](https://cloud.githubusercontent.com/assets/191598/9951510/2e3226fe-5d95-11e5-8a15-7fc82738855f.png)

shaunandrews commented:
Should/could I do be able to do more with the more menu?

![screen shot 2015-09-18 at 1 10 53 am](https://cloud.githubusercontent.com/assets/191598/9952319/3160e33a-5da2-11e5-828c-e3f20889c310.png)

Having that open a single option seems strange. Should we just replace that "..." button with a "Share" label?

aerych commented:
> Having that open a single option seems strange. Should we just replace that "..." button with a "Share" label?

Normally the menu includes "share" and "block" options. We but we don't show the block option on some feeds like Following, Posts I Like, and user created lists that you're following.  (Interestingly we are showing it on the web for Posts I Like but that seems weird.)  We have a "visit site" option on the web menu. Would it make sense to include that option in the app?

shaunandrews commented:
> We but we don't show the block option on some feeds like Following

Should we? What if a site I'm following suddenly starts showing content I find offensive. On the web at least I can unfollow:

![screen shot 2015-09-18 at 11 18 57 am](https://cloud.githubusercontent.com/assets/191598/9963054/16e6279a-5df7-11e5-814e-8be400a02b36.png)

Can we add a follow/unfollow button there?

> We have a "visit site" option on the web menu. Would it make sense to include that option in the app?

Yea, lets try it.

aerych commented:
@shaunandrews  I've updated the menu to include a "Visit Site" option and when viewing `/read/following` it will show options to unfollow and follow instead of block a site.  I'm sorting the list alphabetically but if we want a different order let me know
![simulator screen shot sep 18 2015 12 30 10 pm](https://cloud.githubusercontent.com/assets/1435271/9966413/6755ccfc-5e02-11e5-9436-3161596fe7e9.png)
![simulator screen shot sep 18 2015 12 30 28 pm](https://cloud.githubusercontent.com/assets/1435271/9966420/6eca273a-5e02-11e5-90ad-d8835cdf4c58.png)

Toggling following displays a confirmation on success.  Its not ideal but I think works until we can build a more [appropriate style prompt](https://github.com/wordpress-mobile/WordPress-iOS/issues/3750).

![simulator screen shot sep 18 2015 12 30 14 pm](https://cloud.githubusercontent.com/assets/1435271/9966428/7b0c11e8-5e02-11e5-9a16-a4c1bfd29648.png)

shaunandrews commented:
> I'm sorting the list alphabetically but if we want a different order let me know

Lets put the follow/unfollow always at the top—furthers away from accidental thumb blunders—and Share always at the bottom.

> Toggling following displays a confirmation on success.

I dig it. It'd be great to have some more immediate feedback while the unfollow/follow action does its thing in the background. Its a little easy to assume that nothing happened—and then boom, a notice. Maybe just an intermediate notice with a spinner?

aerych commented:
@koke @shaunandrews I think everything has been addressed. Ready for another round!

shaunandrews commented:
Feels good to me. Nice stuff.

koke commented:
There's been a few times when I tried to tap the site title to open the site view and it opened the post instead. I think all the site header should be a target and not just the label

koke commented:
> Confirm following a site still works via the search bar.

It works although it's really slow (5-10 seconds) and there's no indication of progress other than the network indicator. When it succeeds, you get a HUD message that says 'Following', but nothing else.

It works, and I think that's out of scope, but I wanted to point out it's not a great flow.

koke commented:
When you tap on 'View Site' from the ellipsis menu, the browser view is pushed instead of shown modally like the rest of the app

koke commented:
From the screenshots on #4298:

![emoji-variants](https://cloud.githubusercontent.com/assets/8739/10039130/3e8b9876-61cf-11e5-979d-97ef414dbdfc.png)

I also noticed the post headers are different (avatar, author name, following status) if you come from following and the site list

koke commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift:
```diff
@@ -396,10 +401,15 @@ import Foundation
-        // tagButton.setTitle(title, forState: .Highlighted)
+        var tag = ""
+        if let rawTag = contentProvider?.primaryTag() {
+            if (rawTag.characters.count > 0) {
```
I think `!rawTag.isEmpty` does the same thing.

koke commented on WordPress/Classes/ViewRelated/Reader/ReaderSiteStreamHeader.swift:
```diff
@@ -29,25 +34,74 @@ import Foundation
-        if topic.following {
+        assert(topic.isKindOfClass(ReaderSiteTopic), "Topic must be a site topic")
+
+        let siteTopic = topic as! ReaderSiteTopic
```
I'm wondering why the forced downcast here. If you are certain that `topic` is a `ReaderSiteTopic`, why isn't that the type of the function's argument?

aerych replied:
`configureHeader` is defined in a protocol and implemented by all the headers. Its downcast for the specific use case.

koke replied:
I'm sure there's a Swiftier way to declare that protocol with associated types, but I can't figure it out :wink:

aerych replied:
Let me know if you do ;P

koke commented on WordPress/Classes/ViewRelated/Reader/ReaderSiteStreamHeader.swift:
```diff
@@ -29,25 +34,74 @@ import Foundation
+
+        let siteTopic = topic as! ReaderSiteTopic
+
+        let url = NSURL(string: siteTopic.siteBlavatar)!
```
It's unlikely, but `NSURL.init?(string:)` can fail. If that happens we should show the default avatar instead of crashing :wink:

aerych replied:
I'm all for paranioa. ;)
I can see an issue initing an `NSURL` if `siteTopic.siteBlavatar` is nil (tho it should never be), but to be clear, were you thinking of that or something different?

koke replied:
It's not just `nil`, the string can be an invalid URL. Some examples that return `nil`:

```
NSURL(string: "http://foo.com\\")
NSURL(string: "https://secure.gravatar.com/123a{")
NSURL(string: "ñ")
```

aerych replied:
I wasn't worried about a nil URL since the placeholder would be assigned in those cases anyway (tho a nil string would cause an exception creating the URL).  However, I see that AFNetworking has flagged the argument as non nil so let's check for that too.

koke commented on WordPress/Classes/ViewRelated/Reader/ReaderSiteStreamHeader.swift:
```diff
@@ -29,25 +34,74 @@ import Foundation
+
+    func formattedFollowerCountForTopic(topic:ReaderSiteTopic) -> String {
+        let numberFormatter = NSNumberFormatter()
+        numberFormatter.groupingSeparator = NSLocale.currentLocale().objectForKey(NSLocaleGroupingSeparator) as! String
```
I think you can skip this line as it is the default behavior

<img width="737" alt="screen shot 2015-09-23 at 11 01 21" src="https://cloud.githubusercontent.com/assets/8739/10041169/80f6df3c-61e2-11e5-8f18-1f9fe8c9c000.png">

aerych replied:
Without it I wasn't seeing commas.

aerych replied:
But interestingly now I am...

koke commented on WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:
```diff
@@ -861,12 +1000,12 @@ import Foundation
+
+        let controller = ReaderStreamViewController.controllerWithTagSlug(post.primaryTagSlug)
+        navigationController?.pushViewController(controller, animated: true)
+        // TODO: Analytics
```
Did you forget about this, or is it left for a future PR?

aerych replied:
Future PR :)  Its waiting on this one to be merged.

koke commented on WordPress/Classes/Models/ReaderPost.h:
```diff
@@ -15,6 +15,7 @@ extern NSString * const ReaderPostStoredCommentTextKey;
 @property (nonatomic, strong) NSString *authorDisplayName;
 @property (nonatomic, strong) NSString *authorEmail;
 @property (nonatomic, strong) NSString *authorURL;
+@property (nonatomic, strong) NSString *blavatar;
```
Can I suggest calling this `siteIcon` (or better `siteIconURL`)? Not all the icons are blavatars anymore, the REST API uses that terminology, and in general things are moving towards "site icon"

aerych replied:
I can dig it.

koke commented on WordPress/WordPressApi/WordPressComApi.m:
```diff
@@ -146,6 +146,11 @@ - (WPJSONRequestOperation *)POST:(NSString *)URLString
+    // Skip asserts for non API calls
+    if ([URLString rangeOfString:WordPressComApiClientEndpointURL].location == NSNotFound) {
+        return;
+    }
```
I'm curious: in which scenarios is `WordPressComApi` used for non API calls?

aerych replied:
On example is when following a site entered via the menu's search bar. A HEAD request is made to ensure the site isn't a 404.  It uses the API as a matter of convenience.  The above conditional was added in the 5.6.1 branch, but is also reflected here.

koke commented:
Done reviewing. Looking pretty good in general :smile:

koke commented on WordPress/Classes/ViewRelated/Reader/ReaderSiteStreamHeader.swift:
```diff
@@ -29,25 +34,89 @@ import Foundation
+        } else {
+            avatarImageView.image = placeholder
+        }
+    }
```
Code looks correct, but I just wanted to share that I'm lately enjoying map/flatMap for optional handling. Like this:

```swift
    func configureHeaderImage(siteBlavatar: String?) {
        let placeholder = UIImage(named: defaultBlavatar)

        if let url: NSURL = siteBlavatar.flatMap({ NSURL(string: $0) }) {
            avatarImageView.setImageWithURL(url, placeholderImage: placeholder)
        } else {
            avatarImageView.image = placeholder
        }
    }
```

aerych commented:
Updated with your feedback @koke (plus an issue previewing feeds I noticed). Ready for another round!

koke commented:
:shipit: :tada:

aerych commented:
Thanks @koke & @shaunandrews :)
